--- a/arch/arm64/boot/dts/mediatek/mt7981b.dtsi
+++ b/arch/arm64/boot/dts/mediatek/mt7981b.dtsi
@@ -551,7 +551,12 @@
 			reg = <0 0x15100000 0 0x80000>;
 			resets = <&ethsys 0>;
 			reset-names = "mtketh";
-			status = "disabled";
+			status = "okay";
+			mtketh-wan = "eth1";
+			mtketh-lan = "lan";
+			mtketh-max-gmac = <2>;
+			mtketh-ppe-num = <2>;
+			ext-devices-prefix = "wwan", "usb", "eth";
 		};
 
 		ap2woccif: ap2woccif@151a5000 {
