--- a/arch/arm64/boot/dts/mediatek/mt7981b.dtsi
+++ b/arch/arm64/boot/dts/mediatek/mt7981b.dtsi
@@ -99,18 +99,23 @@
 		};
 
 		wmcpu_emi: wmcpu-reserved@47c80000 {
+			compatible = "mediatek,wmcpu-reserved";
 			reg = <0 0x47c80000 0 0x100000>;
 			no-map;
 		};
 
 		wo_emi0: wo-emi@47d80000 {
+			compatible = "mediatek,wocpu0_emi";
 			reg = <0 0x47d80000 0 0x40000>;
 			no-map;
+			shared = <0>;
 		};
 
 		wo_data: wo-data@47dc0000 {
+			compatible = "mediatek,wocpu_data";
 			reg = <0 0x47dc0000 0 0x240000>;
 			no-map;
+			shared = <1>;
 		};
 	};
 
@@ -456,18 +461,15 @@
         };
 
 		wed: wed@15010000 {
-			compatible = "mediatek,mt7981-wed",
-				     "mediatek,mt7986-wed",
-				     "syscon";
-			reg = <0 0x15010000 0 0x1000>;
+			compatible = "mediatek,wed";
+			wed_num = <2>;
+			/* add this property for wed get the pci slot number. */
+			pci_slot_map = <0>, <1>;
+			reg = <0 0x15010000 0 0x1000>,
+				<0 0x15011000 0 0x1000>;
 			interrupt-parent = <&gic>;
-			interrupts = <GIC_SPI 205 IRQ_TYPE_LEVEL_HIGH>;
-			memory-region = <&wo_emi0>, <&wo_data>;
-			memory-region-names = "wo-emi", "wo-data";
-			mediatek,wo-ccif = <&wo_ccif0>;
-			mediatek,wo-ilm = <&wo_ilm0>;
-			mediatek,wo-dlm = <&wo_dlm0>;
-			mediatek,wo-cpuboot = <&wo_cpuboot>;
+			interrupts = <GIC_SPI 205 IRQ_TYPE_LEVEL_HIGH>,
+					<GIC_SPI 206 IRQ_TYPE_LEVEL_HIGH>;
 		};
 
 		eth: ethernet@15100000 {
@@ -579,27 +581,24 @@
 		};
 
 		wo_dlm0: syscon@151e8000 {
-			compatible = "mediatek,mt7986-wo-dlm", "syscon";
-			reg = <0 0x151e8000 0 0x2000>;
+			compatible = "mediatek,wocpu_dlm",
+						 "mediatek,mt7986-wo-dlm", "syscon";
+			reg = <0 0x151e8000 0 0x2000>,
+				  <0 0x151f8000 0 0x2000>;
 		};
 
 		wo_ilm0: syscon@151e0000 {
-			compatible = "mediatek,mt7986-wo-ilm", "syscon";
+			compatible = "mediatek,wocpu0_ilm",
+						 "mediatek,mt7986-wo-ilm", "syscon";
 			reg = <0 0x151e0000 0 0x8000>;
 		};
 
 		wo_cpuboot: syscon@15194000 {
-			compatible = "mediatek,mt7986-wo-cpuboot", "syscon";
+			compatible = "mediatek,wocpu_boot",
+						 "mediatek,mt7986-wo-cpuboot", "syscon";
 			reg = <0 0x15194000 0 0x1000>;
 		};
 
-		wo_ccif0: syscon@151a5000 {
-			compatible = "mediatek,mt7986-wo-ccif", "syscon";
-			reg = <0 0x151a5000 0 0x1000>;
-			interrupt-parent = <&gic>;
-			interrupts = <GIC_SPI 211 IRQ_TYPE_LEVEL_HIGH>;
-		};
-
 		sgmiisys0: syscon@10060000 {
 			compatible = "mediatek,mt7981-sgmiisys_0", "mediatek,mt7986-sgmiisys_0", "syscon";
 			reg = <0 0x10060000 0 0x1000>;
@@ -663,6 +662,10 @@
 			compatible = "mediatek,mt7981-consys";
 			reg = <0 0x10000000 0 0x8600000>;
 			memory-region = <&wmcpu_emi>;
+			/* This is synced from driver changes */
+			clocks = <&topckgen CLK_TOP_NETSYS_MCU_SEL>,
+				 <&topckgen CLK_TOP_AP2CNN_HOST_SEL>;
+			clock-names = "mcu", "ap2conn";
 		};
 
 		xhci: usb@11200000 {
@@ -749,7 +752,8 @@
 		};
 
 		wifi: wifi@18000000 {
-			compatible = "mediatek,mt7981-wmac";
+			compatible = "mediatek,wbsys",
+						 "mediatek,mt7981-wmac";
 			pinctrl-0 = <&wifi_dbdc_pins>;
 			pinctrl-names = "dbdc";
 			reg = <0 0x18000000 0 0x1000000>,
@@ -764,6 +768,7 @@
 			clock-names = "mcu", "ap2conn";
 			resets = <&watchdog MT7986_TOPRGU_CONSYS_SW_RST>;
 			reset-names = "consys";
+			chip_id = <0x7981>;
 			memory-region = <&wmcpu_emi>;
 			status = "disabled";
 		};
