--- a/arch/arm64/boot/dts/mediatek/mt7986a.dtsi
+++ b/arch/arm64/boot/dts/mediatek/mt7986a.dtsi
@@ -599,7 +599,12 @@
 			resets = <&ethsys 0>;
 			reset-names = "mtketh";
 			mtketh-soc = <&eth>;
-			status = "disabled";
+			status = "okay";
+			mtketh-wan = "eth1";
+			mtketh-lan = "lan";
+			mtketh-max-gmac = <2>;
+			mtketh-ppe-num = <2>;
+			ext-devices-prefix = "wwan", "usb", "eth";
 		};
 
 		wo_ccif0: syscon@151a5000 {
