--- a/drivers/net/ethernet/mediatek/mtk_hnat/hnat.h
+++ b/drivers/net/ethernet/mediatek/mtk_hnat/hnat.h
@@ -229,12 +229,20 @@
 /* If user wants to change default FOE entry number, both DEF_ETRY_NUM and
  * DEF_ETRY_NUM_CFG need to be modified.
  */
-#define DEF_ETRY_NUM		8192
+#if defined(CONFIG_MEDIATEK_NETSYS_V3)
+#define DEF_ETRY_NUM		32768
 /* feasible values : 32768, 16384, 8192, 4096, 2048, 1024 */
-#define DEF_ETRY_NUM_CFG	TABLE_8K
+#define DEF_ETRY_NUM_CFG	TABLE_32K
 /* corresponding values : TABLE_32K, TABLE_16K, TABLE_8K, TABLE_4K, TABLE_2K,
  * TABLE_1K
  */
+#elif defined(CONFIG_MEDIATEK_NETSYS_V2)
+#define DEF_ETRY_NUM		16384
+#define DEF_ETRY_NUM_CFG	TABLE_16K
+#else
+#define DEF_ETRY_NUM		8192
+#define DEF_ETRY_NUM_CFG	TABLE_8K
+#endif
 /* default binding threshold: 30 packets per second */
 #define DEF_BIND_THRESHOLD	30
 #define DEF_TCP_DLTA		7
