--- a/arch/arm64/boot/dts/mediatek/mt7986a.dtsi
+++ b/arch/arm64/boot/dts/mediatek/mt7986a.dtsi
@@ -598,7 +598,12 @@
 			reg = <0 0x15100000 0 0x80000>;
 			resets = <&ethsys 0>;
 			reset-names = "mtketh";
-			status = "disabled";
+			status = "okay";
+			mtketh-wan = "eth1";
+			mtketh-lan = "lan";
+			mtketh-max-gmac = <2>;
+			mtketh-ppe-num = <2>;
+			ext-devices-prefix = "wwan", "usb", "eth";
 		};
 
 		wo_ccif0: syscon@151a5000 {
