--- a/arch/arm64/boot/dts/mediatek/mt7986a.dtsi
+++ b/arch/arm64/boot/dts/mediatek/mt7986a.dtsi
@@ -100,23 +100,30 @@
 		};
 
 		wmcpu_emi: wmcpu-reserved@4fc00000 {
+			compatible = "mediatek,wmcpu-reserved";
 			no-map;
 			reg = <0 0x4fc00000 0 0x00100000>;
 		};
 
 		wo_emi0: wo-emi@4fd00000 {
+			compatible = "mediatek,wocpu0_emi";
 			reg = <0 0x4fd00000 0 0x40000>;
 			no-map;
+			shared = <0>;
 		};
 
 		wo_emi1: wo-emi@4fd40000 {
+			compatible = "mediatek,wocpu1_emi";
 			reg = <0 0x4fd40000 0 0x40000>;
 			no-map;
+			shared = <0>;
 		};
 
 		wo_data: wo-data@4fd80000 {
+			compatible = "mediatek,wocpu_data";
 			reg = <0 0x4fd80000 0 0x240000>;
 			no-map;
+			shared = <1>;
 		};
 	};
 
@@ -139,6 +146,16 @@
 			#interrupt-cells = <3>;
 		};
 
+		consys: consys@10000000 {
+			compatible = "mediatek,mt7986-consys";
+			reg = <0 0x10000000 0 0x8600000>;
+			memory-region = <&wmcpu_emi>;
+			/* This is synced from driver changes */
+			clocks = <&topckgen CLK_TOP_CONN_MCUSYS_SEL>, /* DIFFERENT from mt7981 */
+				 <&topckgen CLK_TOP_AP2CNN_HOST_SEL>;
+			clock-names = "mcu", "ap2conn";
+		};
+
 		infracfg: infracfg@10001000 {
 			compatible = "mediatek,mt7986-infracfg", "syscon";
 			reg = <0 0x10001000 0 0x1000>;
@@ -147,7 +164,8 @@
 		};
 
 		wed_pcie: wed-pcie@10003000 {
-			compatible = "mediatek,mt7986-wed-pcie",
+			compatible = "mediatek,wed_pcie",
+						 "mediatek,mt7986-wed-pcie",
 				     "syscon";
 			reg = <0 0x10003000 0 0x10>;
 		};
@@ -530,32 +548,26 @@
 			};
         };
 
-		wed0: wed@15010000 {
-			compatible = "mediatek,mt7986-wed",
-				     "syscon";
-			reg = <0 0x15010000 0 0x1000>;
+		wed: wed@15010000 {
+			compatible = "mediatek,wed";
+			wed_num = <2>;
+			/* add this property for wed get the pci slot number. */
+			pci_slot_map = <0>, <1>;
+			reg = <0 0x15010000 0 0x1000>,
+				<0 0x15011000 0 0x1000>;
 			interrupt-parent = <&gic>;
-			interrupts = <GIC_SPI 205 IRQ_TYPE_LEVEL_HIGH>;
-			memory-region = <&wo_emi0>, <&wo_data>;
-			memory-region-names = "wo-emi", "wo-data";
-			mediatek,wo-ccif = <&wo_ccif0>;
-			mediatek,wo-ilm = <&wo_ilm0>;
-			mediatek,wo-dlm = <&wo_dlm0>;
-			mediatek,wo-cpuboot = <&wo_cpuboot>;
+			interrupts = <GIC_SPI 205 IRQ_TYPE_LEVEL_HIGH>,
+					<GIC_SPI 206 IRQ_TYPE_LEVEL_HIGH>;
 		};
 
-		wed1: wed@15011000 {
-			compatible = "mediatek,mt7986-wed",
-				     "syscon";
-			reg = <0 0x15011000 0 0x1000>;
+		wed2: wed2@15011000 {
+			compatible = "mediatek,wed2";
+			wed_num = <2>;
+			reg = <0 0x15010000 0 0x1000>,
+				<0 0x15011000 0 0x1000>;
 			interrupt-parent = <&gic>;
-			interrupts = <GIC_SPI 206 IRQ_TYPE_LEVEL_HIGH>;
-			memory-region = <&wo_emi1>, <&wo_data>;
-			memory-region-names = "wo-emi", "wo-data";
-			mediatek,wo-ccif = <&wo_ccif1>;
-			mediatek,wo-ilm = <&wo_ilm1>;
-			mediatek,wo-dlm = <&wo_dlm1>;
-			mediatek,wo-cpuboot = <&wo_cpuboot>;
+			interrupts = <GIC_SPI 205 IRQ_TYPE_LEVEL_HIGH>,
+					<GIC_SPI 206 IRQ_TYPE_LEVEL_HIGH>;
 		};
 
 		eth: ethernet@15100000 {
@@ -599,7 +611,7 @@
 			mediatek,ethsys = <&ethsys>;
 			mediatek,sgmiisys = <&sgmiisys0>, <&sgmiisys1>;
 			mediatek,wed-pcie = <&wed_pcie>;
-			mediatek,wed = <&wed0>, <&wed1>;
+			mediatek,wed = <&wed>, <&wed2>;
 			status = "disabled";
 		};
 
@@ -617,47 +629,51 @@
 			ext-devices-prefix = "wwan", "usb", "eth";
 		};
 
-		wo_ccif0: syscon@151a5000 {
-			compatible = "mediatek,mt7986-wo-ccif", "syscon";
-			reg = <0 0x151a5000 0 0x1000>;
-			interrupt-parent = <&gic>;
-			interrupts = <GIC_SPI 211 IRQ_TYPE_LEVEL_HIGH>;
-		};
-
-		wo_ccif1: syscon@151ad000 {
-			compatible = "mediatek,mt7986-wo-ccif", "syscon";
-			reg = <0 0x151ad000 0 0x1000>;
+		wdma: wdma@15104800 {
+			compatible = "mediatek,wed-wdma";
+			reg = <0 0x15104800 0 0x400>,
+				<0 0x15104c00 0 0x400>;
+ 		};
+
+		ap2woccif: ap2woccif@151a5000 {
+			compatible = "mediatek,ap2woccif";
+			reg = <0 0x151a5000 0 0x1000>,
+				<0 0x151ad000 0 0x1000>;
 			interrupt-parent = <&gic>;
-			interrupts = <GIC_SPI 212 IRQ_TYPE_LEVEL_HIGH>;
+			interrupts = <GIC_SPI 211 IRQ_TYPE_LEVEL_HIGH>,
+					<GIC_SPI 212 IRQ_TYPE_LEVEL_HIGH>;
 		};
 
 		wo_ilm0: syscon@151e0000 {
-			compatible = "mediatek,mt7986-wo-ilm", "syscon";
+			compatible = "mediatek,wocpu0_ilm",
+						 "mediatek,mt7986-wo-ilm", "syscon";
 			reg = <0 0x151e0000 0 0x8000>;
 		};
 
 		wo_ilm1: syscon@151f0000 {
-			compatible = "mediatek,mt7986-wo-ilm", "syscon";
+			compatible = "mediatek,wocpu1_ilm",
+						 "mediatek,mt7986-wo-ilm", "syscon";
 			reg = <0 0x151f0000 0 0x8000>;
 		};
 
 		wo_dlm0: syscon@151e8000 {
-			compatible = "mediatek,mt7986-wo-dlm", "syscon";
-			reg = <0 0x151e8000 0 0x2000>;
-		};
-
-		wo_dlm1: syscon@151f8000 {
-			compatible = "mediatek,mt7986-wo-dlm", "syscon";
-			reg = <0 0x151f8000 0 0x2000>;
+			compatible = "mediatek,wocpu_dlm",
+						 "mediatek,mt7986-wo-dlm", "syscon";
+			reg = <0 0x151e8000 0 0x2000>,
+				  <0 0x151f8000 0 0x2000>;
+			resets = <&ethsysrst 0>;
+			reset-names = "wocpu_rst";
 		};
 
 		wo_cpuboot: syscon@15194000 {
-			compatible = "mediatek,mt7986-wo-cpuboot", "syscon";
+			compatible = "mediatek,wocpu_boot",
+						 "mediatek,mt7986-wo-cpuboot", "syscon";
 			reg = <0 0x15194000 0 0x1000>;
 		};
 
 		wifi: wifi@18000000 {
-			compatible = "mediatek,mt7986-wmac";
+			compatible = "mediatek,wbsys",
+						 "mediatek,mt7986-wmac";
 			reg = <0 0x18000000 0 0x1000000>,
 			      <0 0x10003000 0 0x1000>,
 			      <0 0x11d10000 0 0x1000>;
@@ -670,6 +686,7 @@
 				     <GIC_SPI 214 IRQ_TYPE_LEVEL_HIGH>,
 				     <GIC_SPI 215 IRQ_TYPE_LEVEL_HIGH>,
 				     <GIC_SPI 216 IRQ_TYPE_LEVEL_HIGH>;
+			chip_id = <0x7986>;
 			memory-region = <&wmcpu_emi>;
 		};
 
